### PR TITLE
docs: full technical documentation suite indexed in docs/README.md

### LINKED ISSUE #131: Documentation suite: one technical document per pipeline surface, indexed in docs/README.md
## Problem

`docs/` currently holds only two documents (`GPU_RUNTIME_GUIDE.md`, `CONFIG_AND_CLI.md`) plus a `placeholder` file whose TODO lists everything still missing:

> Document (individually) db architecture, logger architecture, resource lifecycle manager, resource monitoring, training orchestration, inference orchestration, preprocessing (including energy detection and signal injection), model architecture design (beta-VAE, RF), github actions (especially the claude integrations)

The v1 pipeline surfaces (memmap round data + producer overlap, stage-machine training retries, latent traversal, vectorized energy detection, PFB equalization, per-cadence streaming inference + manifest retries, the inference visualization suite, DB schema migrations + supersede semantics) have detailed docstrings but no long-form documentation tying them together. New contributors and future maintainers have to reverse-engineer design intent from the code.

## Proposal

Replace `docs/placeholder` with a full documentation suite — one technical document per logical pipeline surface, written against the current code with file/symbol references, indexed in `docs/README.md`:

- `docs/ARCHITECTURE.md` — system overview, data model (cadence/stamp/snippet), module map + mermaid data-flow diagram, init order, singleton rationale, directory/tag conventions, artifact locations
- `docs/TRAINING_PIPELINE.md` — curriculum schedules, round lifecycle, memmap round data + producer overlap, distributed strategy + gradient accumulation, adaptive LR, run-state manifest + retry semantics, every training plot explained, latent traversal
- `docs/INFERENCE_PIPELINE.md` — CSV catalog format + cadence grouping, per-cadence streaming flow, model resolution, threshold semantics, per-cadence manifest + retry, every inference artifact/plot explained
- `docs/PREPROCESSING.md` — energy detection end-to-end (DC spike, spline vs PFB bandpass math, vectorized D'Agostino-Pearson derivation, dedup, overlap stamps, downsample-at-extraction, storage math), signal injection, training background loading
- `docs/MODELS.md` — beta-VAE architecture tables, composite loss math incl. clustering terms, RF feature layout, threshold/confidence semantics, SHAP artifact meanings
- `docs/DATABASE.md` — schema per table, writer-thread design, flush protocol, migration mechanism, query API + supersede filtering, growth expectations
- `docs/RUNTIME_SERVICES.md` — logger (queue architecture, Slack threading, stderr redirect gotcha), ResourceManager (pool/SHM/process lifecycle, signal handling, cleanup order), monitor (metrics, PSS rationale)
- `docs/TESTING.md` — suite layout, markers, fixtures/singleton resets, CI, cluster-marked tests, adding-tests checklist
- `docs/BENCHMARKING.md` — stage timers, report tool usage, micro-benchmarks, baseline numbers (once the benchmarking suite lands)
- `docs/RELEASE.md` — release runbook, version-coupling contract (PyPI / GitHub / HF), CD workflow internals, trusted-publishing setup
- `docs/GITHUB_AUTOMATION.md` — every workflow in `.github/workflows/` (triggers, behavior, dedup markers, the assistant-handle rule), issue/PR lifecycle

Plus: update the `docs/README.md` index table, keep `GPU_RUNTIME_GUIDE.md` / `CONFIG_AND_CLI.md` with refreshed cross-references, delete `docs/placeholder`, and refresh the doc lists in `CLAUDE.md`, the repo-context skill, and `CONTRIBUTING.md`'s structure tree.

## Verification

No cluster needed: verify every intra-repo link resolves (scripted check), markdown renders sanely (tables/mermaid), and code snippets match current signatures.

### PR DESCRIPTION
**Pull request by zachtheyek**
Closes #131

## Stacking

Based on `integration/pr11-base` (the all-code base: training pipeline + retry overhaul, inference energy-detection + PFB + retry/viz, benchmarking suite #134, HuggingFace integration #130, release engineering #133). Being docs-only, this PR **documents everything before it** and carries no code changes. Merge after the code PRs it describes.

## What this adds

The full `docs/` technical suite — one document per pipeline surface — indexed in `docs/README.md`, replacing `docs/placeholder`:

| Document | Surface |
|---|---|
| `ARCHITECTURE.md` | System map, data model, module/data-flow, init order, singletons, artifact layout |
| `TRAINING_PIPELINE.md` | Curriculum, round lifecycle, memmap round data + producer, retries, every training plot |
| `INFERENCE_PIPELINE.md` | Catalog grouping, streaming flow, **HF model resolution**, tag guards, manifest retries, viz |
| `PREPROCESSING.md` | Energy detection math, spline vs PFB, vectorized D'Agostino–Pearson, dedup, downsample-at-extraction, injection |
| `MODELS.md` | Beta-VAE architecture + composite loss, RF feature layout, threshold/SHAP semantics |
| `DATABASE.md` | Every table incl. **`pipeline_stages`**, writer thread, flush/supersede, **migration mechanism** |
| `RUNTIME_SERVICES.md` | Logger/Slack, ResourceManager lifecycle, monitor + **stage annotations** |
| `TESTING.md` | Suite layout, markers, isolation fixtures, CI, cluster smokes |
| `BENCHMARKING.md` | **New** — stage timers, `pipeline_stages`, report tool, micro-benchmarks, baselines |
| `RELEASE.md` | Version-coupling contract, CD gates (verify-don't-create), trusted publishing, runbook |
| `GITHUB_AUTOMATION.md` | Workflow catalog, dedup markers, assistant-handle rules |

Also refreshes `CLAUDE.md`, the repo-context skill, and `CONTRIBUTING.md`, and extends `CONFIG_AND_CLI.md` / `GPU_RUNTIME_GUIDE.md` cross-references.

### Phase B (this session) — the deferred, code-dependent pieces

Written against the landed modules (`benchmark.py`, `hf_hub.py`, `tag_guards.py`, `db/db.py`, `utils/benchmark_report.py`, `benchmarks/`, `.github/workflows/release.yml`, `utils/hf_tag_release.py`):

- **`docs/BENCHMARKING.md` (new)** — `stage_timer`/`record_stage` with thread-local dot-name nesting, the instrumented stage map, the `pipeline_stages` table (schema v3), `monitor.annotate_stages` overlay, `benchmark_report.py` (stage tree / flame PNG / suggestion rules), the `benchmarks/` micro-benchmarks, and baseline numbers (123× vectorized normality and 11.9× PFB-vs-spline on blpc3; end-to-end inference ~6.2× and ED ~45 min → ~2 min/cadence; training RSS held ~26% vs the old round-2 OOM).
- **`INFERENCE_PIPELINE.md`** — filled the `<!-- PHASE-B: HF model resolution -->` gap: the full artifact-resolution chain (explicit paths → `--hf-revision` → installed-release `v{__version__}` with the strict dev-version guard → latest release/`final_vX` tag), the all-or-none validation, and the inference-scope tag-dedup guards. Corrected the now-stale "all three paths required" text.
- **`DATABASE.md`** — added the `pipeline_stages` table and a `user_version` → feature map (v1 superseded columns, v2 `inference_cadences`, v3 `pipeline_stages`), verified against `_migrate_schema`; bumped the version/table-count references.
- **`RUNTIME_SERVICES.md`** — documented the `monitor.annotate_stages` CPU-panel overlay (config knob, depth≤2 rule, flush-first).
- **`docs/README.md`** — real `BENCHMARKING.md` index row replacing the placeholder comment.
- **`RELEASE.md`** — cross-checked every CD gate against the implemented `release.yml`/`hf_tag_release.py`; added the build-stage wheel smoke and the GitHub Release artifact attachment for fidelity.

## Verification

Docs-only; no runtime surface. Checks performed:

- **Link + anchor check** (scripted: every relative markdown link and `#anchor` across all docs resolved against the filesystem / target headings):
  ```
  Checked 243 relative links across 17 markdown files.
  All relative links and anchors resolve. 0 broken.
  ```
- **Code fidelity** — every cited symbol/flag confirmed against source: `stage_timer`/`record_stage`/`current_stage`/`_ACTIVE_STAGES` (`benchmark.py`), `write_pipeline_stage`/`query_pipeline_stages`/`_SCHEMA_VERSION=3`/`_migrate_schema` (`db/db.py`), `select_annotation_spans`/`_ANNOTATION_MAX_DEPTH`/`annotate_stages` (`monitor.py` + config default), `build_stage_tree`/`top_slowest`/suggestion thresholds (`benchmark_report.py`), `version_default_revision`/`_SEMVER_TAG_PATTERN`/`resolve_hf_revision`/`resolve_inference_artifacts` (`hf_hub.py`), `enforce_tag_guards`/`find_inference_tag_collisions`/`--force-tag` (`tag_guards.py`), and the release-gate ordering in `release.yml`.
- **`pre-commit run --all-files`** — green.
- Unit suite unaffected (no test fixture reads `docs/`; CLAUDE.md/CONTRIBUTING structural fixtures untouched by content-only edits). CI will exercise it.
- Both commits carry verified GPG signatures.

Note for a follow-up (not in scope here): `utils/benchmark_report.py`'s encode-input-bound suggestion prints `--inf-batch-size`, but the real flag is `--per-replica-batch-size`; `BENCHMARKING.md` documents the correct flag.

## AI assistance

Authored with Claude Code (Anthropic) — Phase B doc drafting against the landed code, plus the scripted link/anchor check and code-symbol verification. Reviewed for factual accuracy against the source. Disclosed per `AI_POLICY.md`.

### CODE REVIEW

**Comment by zachtheyek** [comment]
Applied the review polish (`3b2b044`, GPG-signed):

1. **ARCHITECTURE.md** — moved `setup_gpu_strategy()` into prose so the `main.py` link no longer renders a stray trailing backtick span.
2. **DATABASE.md** — made the `~24 rows/cadence` derivation explicit: 18 intensity rows (6 stats × 3 stages A/B/C) + mean 6 signal-characteristic rows across the 0/6/6/12 equal-weighted types.
3. **GITHUB_AUTOMATION.md** — added a callout unpacking the dense `claude-update-docs` row (deterministic shell regen → handle-mention issue → assistant PR, split because the follow-up run has no `python` in its allowlist).
4. **RUNTIME_SERVICES.md** — verified the KNOWN_ISSUES.md references against the current file: #6 = Resource Monitor CPU Undercounting, #10 = Missing Sections in Resource Monitor Plots, #11 = Negative Memory Freed Calculation. No drift; no change needed.
5. The `benchmark_report.py` `--inf-batch-size` string is being fixed in #134 by another agent; this doc already documents the correct flag (`--per-replica-batch-size`), so no doc change.

Verification: `pre-commit run` green on the touched files; scripted link+anchor check — **Checked 243 relative links across 17 markdown files. All resolve. 0 broken.**

**Comment by zachtheyek** [comment]
@claude please perform a second-pass code review of this PR. Re-review the current diff for any remaining correctness bugs, security issues, or significant quality problems, and post your findings. If the diff is clean, please say so explicitly. (Final verification round before merge.)

**Comment by zachtheyek** [comment]
Pushed a docs-only accuracy pass to `docs/TESTING.md` (commit 272bef1):

1. **Completed the `## Layout` unit-test tree.** Added the four `tests/unit/` files that exist on this branch but were missing from the fenced tree — `test_tag_guards.py`, `test_hf_hub.py`, `test_benchmark.py`, `test_logger.py` — placed near their related domains, with terse one-line descriptions read off each file and column alignment / the `└──` last-entry connector preserved. (`test_monitor.py` stays correctly absent — none exists.)

2. **Added a `## Coverage and deliberate gaps` section** so a reader can tell deliberate gap from oversight:
   - `monitor` has no dedicated unit-test module by design — it's dominated by the 1 Hz background sampling thread (PSS process-tree stats) and matplotlib rendering, both low-value to unit-test; behavior is covered by the integration smokes and by manual inspection of the resource-utilization plot uploaded to Slack. (Its one pure helper, `select_annotation_spans`, is covered in `test_benchmark.py`.)
   - `logger` is unit-tested only for the `StreamToLogger` stdout/stderr-redirect probes (`isatty`/`writable`/`readable`/`fileno`) in `test_logger.py`; QueueListener, SlackHandler, and the stderr-to-logger redirect are exercised by the integration smokes.
   - One line noting inference logic is unit-tested at the function level with end-to-end behavior in `test_inference_smoke.py`.

Every module/behavior cited was verified against the source on this branch. Relative links in the file still resolve (0 broken) and `pre-commit run --all-files` passes.

**Comment by zachtheyek** [comment]
Pushed two small docs fixes (f77cbd3):

1. **`SKILL.md` — stale placeholder text.** Updated the `# TEST:` grep-marker description; it no longer says "no suite yet" and now reads `(behavior to verify — now backed by the `tests/` suite)`. Grepped the rest of the file for other placeholder-era testing wording; this was the only occurrence.

2. **`docs/TESTING.md` — slack_handler coverage gap.** Extended the "Coverage and deliberate gaps" `logger` bullet to explicitly call out `src/aetherscan/logger/slack_handler.py`: it has no unit tests despite carrying real logic (batching, level-to-color coding, exponential-backoff retries, consecutive-failure cooldown throttling, image uploads) — all exercised only by the integration smokes. Also noted its two known `# BUG:` markers (batch color-by-priority, long-message truncation) as observability-only cosmetics that do not affect pipeline results. Descriptions were verified against the module on this branch.

Verification: relative-link/anchor check over `docs/` passes (0 broken), `pre-commit run --all-files` passes, and the only changed files are `.claude/skills/aetherscan-repo-context/SKILL.md` and `docs/TESTING.md`.

**Comment by zachtheyek** [comment]
Documented the **automated-review response loop for agents** (commit d9e52cc).

- **`CLAUDE.md`** — new terse `## Code review` subsection: wait for the automated `claude-code-review` first pass, validate each note against your own read of the code, fix genuine gaps in self-contained commits on the same PR, rebut wrong notes concretely, post one consolidated reply, and re-invoke via the intentional-handle case; loop until clean/LGTM or the review drifts off-theme / turns nonsensical, then stop and explain why.
- **`.claude/skills/aetherscan-repo-context/SKILL.md`** — fuller paragraph in the Contribution Workflow section spelling out the same loop, complementing (not duplicating) the existing "Invoking vs. mentioning the assistant" guidance and pointing at `docs/GITHUB_AUTOMATION.md`.

Both describe the assistant handle using the repo's existing convention rather than the live trigger string. Docs-only; `pre-commit run --all-files` passes, only these two files changed, GPG-signed.

**Comment by zachtheyek** [comment]
Added a dedicated **## Testing** section to the on-demand `SKILL.md` (placed between Code Style & Conventions and Contribution Workflow), so the skill's testing coverage matches its role as the expanded form of CLAUDE.md instead of being scattered across the Project Structure tree and the comment-markers line.

The new section concisely covers:
- Suite shape: `tests/unit/` (the CPU-only CI surface) vs `tests/integration/` (gpu/cluster-marked end-to-end smokes on a cluster).
- The exact default-selection command CI runs (`pytest -m "not gpu and not cluster" -q`), plus the `pythonpath = ["src"]` note (no `PYTHONPATH=src` prefix for pytest) and `--strict-markers`.
- Marker table (`slow` / `gpu` / `cluster` / `integration`).
- A summary of the autouse `conftest.py` isolation fixture (tmp-path `AETHERSCAN_*`, five-singleton `_reset()`, signal/stream restore) so tests never touch real data or leak state.
- Discipline (run before claiming; ship unit tests with new logic) and the TF-at-collection gotcha (run the subset you can and say what you ran).
- A pointer to `docs/TESTING.md` for the full deep dive.

De-dup: trimmed the `tests/` blurb in the Project Structure tree to a one-liner that cross-references the new section. No overlap with the automated-review loop paragraph already in Contribution Workflow.

Verification: relative-link/anchor check over `docs/` + the touched `SKILL.md` reports 0 broken (the new `docs/TESTING.md` reference resolves); `pre-commit run --all-files` passes; only `SKILL.md` changed. Commit is GPG-signed.

**Comment by zachtheyek** [comment]
Fresh-eyes **documentation drift/coverage review** (6 doc-cluster reviewers + a coverage critic + adversarial verification). The suite predated #170 (dashboard), #129/#134 (schema bumps), and #133 (packaging/CD), so it had real drift. **30 confirmed fixes applied** (`3f6fde2`, `35fa083`); verified against the code, all links resolve, pre-commit clean:

- **Live dashboard (#170) was undocumented across the whole suite** → added to the ARCHITECTURE module map + init sequence, a RUNTIME_SERVICES "Live monitoring dashboard" section, CONFIG_AND_CLI (fields + `--dashboard`/`--dashboard-port`), RELEASE (`[dashboard]` extra), TESTING (`test_dashboard*`), and the CONTRIBUTING/SKILL module trees.
- **DATABASE.md was a schema version behind** → `_SCHEMA_VERSION` 3→4, `pipeline_stages` v3→v4 (+ the anchor-link ripple to BENCHMARKING/RUNTIME_SERVICES), added the v3 `config_fingerprint` ALTER row + the `inference_cadences.config_fingerprint` column.
- **TRAINING_PIPELINE**: `data_gen_task_size` 256→64, added the 6th `hf_upload` stage, fixed the round-data default root.
- **INFERENCE_PIPELINE**: documented the #129 config-fingerprint resume guard; `parallel_coarse_chans`→`coarse_channel_log_interval`.
- **RELEASE**: signed-tag gate as gate 1 (+ SHA-pinned pypa action); corrected the `huggingface_hub` "base-image-implicit" claim.
- **ARCHITECTURE/CONTRIBUTING/SKILL** module maps: added `benchmark`/`hf_hub`/`tag_guards`.
- **§5A cross-wiring**: CLAUDE.md `## Testing` now links `docs/TESTING.md`; PREPROCESSING got the confirmed GBT512 PFB provenance + the `bug=true` caveat (#180); CONTRIBUTING's stale release-workflow TODO replaced.

@claude please review the docs suite for any remaining drift vs the merged code.

**Comment by zachtheyek** [comment]
Thanks — all four applied (`[latest]`), verified against the code:

1. **ARCHITECTURE init sequence**: added `resolve_inference_artifacts(args)` as inference-only step 7 (it runs at main.py:819, between `parse_args` and `apply_saved_config`, writing the HF-resolved cache paths onto `args`) and renumbered the rest 8–14.
2. **ARCHITECTURE singleton list**: added `get_logger()` / `get_monitor()` (logger.py:355 / monitor.py:737) alongside the other three.
3. **CONFIG_AND_CLI tally**: 12 shared (9 Pattern B) → **14 shared (11 Pattern B + 3 Pattern C)** — the two `--dashboard`/`--dashboard-port` flags my earlier pass added weren't reflected in the count.
4. **CONTRIBUTING + SKILL utils trees**: added `benchmark_report.py` + `hf_tag_release.py`.

Numbering re-verified (6→14, no gaps), links resolve, pre-commit clean. @claude one more pass to confirm the suite is drift-free.
